Variant inference in case expressions (#3647)
gleam-lang/gleam: CHANGELOG.md

  the Erlang target if any of their arguments were discarded.
  ([Giacomo Cavalieri](https://github.com/giacomocavalieri))

> - The compiler now narrows the variant of custom types when pattern matching,
>   allowing for safe record updates if the variant is known:

Reviewer: Talking with Louis and Hayleigh at code beam we agreed that we should be really clear in framing this as "safe record updates" instead of _narrowing_ of any type to avoid misleading people into thinking this is type narrowing. This could be reworded as:
```suggestion
- The compiler can now allow you to perform record updates for specific
  variants of a type after pattern matching on it:
```

Author: Sounds good!

- Optimised creation of bit arrays on the JavaScript target.
  ([Richard Viney](https://github.com/richard-viney))

- The compiler can now infer the variant of custom types when pattern matching,
  allowing for safe record updates if the variant is known, or access fields

>   which are only present on one variant:

Reviewer: Instead of cramming it into a single entry I'd split it in two: one for the record updates and one for the record access, so the example can be more focused

Author: Good idea, I wasn't super happy with how the entry looked

Reviewer 2: It'd also need to bring up pattern matching!
